File: java/allaboutthatbase.java
import java.util.ArrayList;
import java.util.Scanner;

public class allaboutthatbase {
	static final char[] allBaseType = new char[] { '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f',
			'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
			'0' };
	static int toBase10(String digits, ArrayList<Character> basedigits) {
		int result = 0;
		for (int i = 0; i < digits.length(); i++)
			result += basedigits.indexOf(digits.charAt(i)) * Math.pow(basedigits.size(), digits.length() - i - 1);
		return result;
	}

	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int N = scan.nextInt();
		scan.nextLine();
		while (N-- > 0) {
			String[] exp = scan.nextLine().split(" ");
			String X = exp[0];
			String op = exp[1];
			String Y = exp[2];
			String Z = exp[4];
			int minBaseType = 1;
			char maxDigit = Character.MIN_VALUE;
			boolean hasZero = false;
			for (String A : new String[] { X, Y, Z })
				for (char c : A.toCharArray()) {
					if (maxDigit < c)
						maxDigit = c;
					if (!hasZero && c == '0')
						hasZero = true;
				}
			if (maxDigit != '1' || hasZero)
				if (Character.isDigit(maxDigit)) {
					int _base = Character.getNumericValue(maxDigit) + 1;
					minBaseType = _base > 1 ? _base : 2;
				} else
					minBaseType = maxDigit - 'a' + 11;
			StringBuilder sb = new StringBuilder();
			for (int currentBaseType = minBaseType; currentBaseType <= 36; currentBaseType++) {
				int result = 0;
				if (currentBaseType == 1) {
					switch (op) {
					case "+":
						result = X.length() + Y.length();
						break;
					case "-":
						result = X.length() - Y.length();
						break;
					case "*":
						result = X.length() * Y.length();
						break;
					default:
						if (Z.length() * Y.length() == X.length())
							sb.append("1");
						break;
					}
					if (Z.length() == result)
						sb.append("1");
				} else {
					if (currentBaseType == 0)
						currentBaseType = 36;
					ArrayList<Character> currentBaseDigits = new ArrayList<>(currentBaseType);
					if (currentBaseType == 1)
						currentBaseDigits.add('1');
					else if (currentBaseType >= 2 && currentBaseType <= 10)
						for (int i = 0; i < currentBaseType; i++)
							currentBaseDigits.add(String.valueOf(i).charAt(0));
					else {
						for (int i = 0; i < 10; i++)
							currentBaseDigits.add(String.valueOf(i).charAt(0));
						for (int i = 0; i < currentBaseType - 10; i++)
							currentBaseDigits.add((char) ('a' + i));
					}
					int X_int = toBase10(X, currentBaseDigits);
					int Y_int = toBase10(Y, currentBaseDigits);
					int Z_int = toBase10(Z, currentBaseDigits);
					switch (op) {
					case "+":
						result = X_int + Y_int;
						break;
					case "-":
						result = X_int - Y_int;
						break;
					case "*":
						result = X_int * Y_int;
						break;
					default:
						if (X_int != Y_int * Z_int)
							continue;
						else
							result = X_int / Y_int;
						break;
					}
					if (result == Z_int)
						sb.append(allBaseType[currentBaseType - 1]);
				}
			}
			System.out.println(sb.length() == 0 ? "Invalid" : sb);
		}
		scan.close();
	}
}

File: ts/src/allaboutthatbase.ts
import readline = require('readline')

function getBases (X: string, op: string, Y: string, Z: string): string[] {
  const bases: string[] = []
  const minBase = getMinimumBase(X, Y, Z)
  for (let base = minBase; base <= maxBase; base++) {
    if (checkBase(X, op, Y, Z, base)) {
      bases.push(mapBaseToString(base))
    }
  }
  return bases
}

function getMinimumBase (X: string, Y: string, Z: string): number {
  const allLetters = X.concat(Y).concat(Z)
  const maxDigit = allLetters.split('').reduce((prev, curr) => {
    return curr > prev ? curr : prev
  })
  if (regLettersOnly.test(maxDigit)) {
    return maxDigit.charCodeAt(0) - startChar + 11
  }
  if (maxDigit === '1' && !allLetters.includes('0')) {
    return 1
  }
  return Math.max(parseInt(maxDigit) + 1, 2)
}

function checkBase (
  X: string,
  op: string,
  Y: string,
  Z: string,
  base: number
): boolean {
  const XInt = toBase10(X, base)
  const YInt = toBase10(Y, base)
  const ZInt = toBase10(Z, base)
  return checkBaseAny(XInt, op, YInt, ZInt)
}

function checkBaseAny (X: number, op: string, Y: number, Z: number): boolean {
  switch (op) {
    case '+':
      return Z === X + Y
    case '-':
      return Z + Y === X
    case '*':
      return Z === X * Y
    default:
      return Z * Y === X
  }
}

function toBase10 (num: string, base: number): number {
  const source = baseTypes.get(mapBaseToString(base)) ?? []
  return base === 1
    ? num.length // base 1
    : num
      .split('')
      .map((n) => source.indexOf(n))
      .reverse()
      .reduce((acc, curr, index) => {
        return acc + curr * Math.pow(source.length, index)
      })
}

function mapBaseToString (base: number): string {
  if (base === maxBase) {
    return '0'
  }
  return base < 10
    ? base.toString()
    : String.fromCodePoint(startChar + base - 10)
}

function generateSourceDigits (base: number): string[] {
  if (base === 1) {
    return ['1']
  }

  const baseDigits: string[] = []
  for (let i = 0; i < Math.min(base, 10); i++) {
    baseDigits.push(i.toString())
  }
  for (let i = 0; i < base - 10; i++) {
    baseDigits.push(String.fromCodePoint(startChar + i))
  }
  return baseDigits
}

function generateBaseTypes (): void {
  for (let i = 1; i <= maxBase; i++) {
    baseTypes.set(mapBaseToString(i), generateSourceDigits(i))
  }
}

let lineCount = 0
const regLettersOnly = /^[a-z]$/
const startChar = 'a'.charCodeAt(0)
const maxBase = 36
const baseTypes = new Map<string, string[]>()

generateBaseTypes()

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    if (lineCount > 0) {
      const [X, op, Y, , Z] = line.split(' ')
      const bases = getBases(X ?? '', op ?? '', Y ?? '', Z ?? '')
      console.log(`${bases.length === 0 ? 'invalid' : bases.join('')}`)
    }
    lineCount++
  })
  .on('close', () => {
    process.exit(0)
  })
